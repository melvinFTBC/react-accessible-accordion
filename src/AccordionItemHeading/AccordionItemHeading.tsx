import { default as classnames } from 'classnames';
import * as React from 'react';
import { UUID } from '../ItemContainer/ItemContainer';
import { getClosestElement } from '../helpers/getClosestElement';

type AccordionItemHeadingProps = React.HTMLAttributes<HTMLDivElement> & {
    expandedClassName: string;
    expanded: boolean;
    uuid: UUID;
    disabled: boolean;
    setExpanded(uuid: UUID, expanded: boolean): void;
};

type AccordionItemHeadingState = {};

export default class AccordionItemHeading extends React.Component<
    AccordionItemHeadingProps,
    AccordionItemHeadingState
> {
    handleClick = (): void => {
        const { uuid, expanded, setExpanded } = this.props;

        setExpanded(uuid, !expanded);
    };

    handleKeyPress = (evt: React.KeyboardEvent<HTMLDivElement>): void => {
        const keyCode = evt.which.toString();
        const parentAccordion = getClosestElement(evt.target, '[data-accordion-component="Accordion"]');
        const accordionItems = parentAccordion.querySelectorAll('[data-accordion-component="AccordionItemHeading"]');

        switch (keyCode) {
            case '13':
            case '32':
                evt.preventDefault();
                this.handleClick();
                break;
            case '36':
                // home
                evt.preventDefault();
                accordionItems[0].focus();
                break;
            case '35':
                // end
                evt.preventDefault();
                accordionItems[accordionItems.length - 1].focus();
                break;
            default:
                break;
        }
    };

    render(): JSX.Element {
        const {
            className,
            expandedClassName,
            setExpanded,
            expanded,
            uuid,
            disabled,
            ...rest
        } = this.props;

        const id = `accordion__heading-${uuid}`;
        const ariaControls = `accordion__panel-${uuid}`;

        const headingClassName = classnames(className, {
            [expandedClassName]: expandedClassName && expanded,
        });

        return (
            <div
                id={id}
                aria-expanded={expanded}
                aria-controls={ariaControls}
                className={headingClassName}
                aria-disabled={disabled}
                onClick={this.handleClick}
                role="button"
                tabIndex={0}
                data-accordion-component="AccordionItemHeading"
                onKeyDown={this.handleKeyPress}
                {...rest}
            />
        );
    }
}