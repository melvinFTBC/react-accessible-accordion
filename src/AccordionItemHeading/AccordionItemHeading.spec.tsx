import { mount, ReactWrapper } from 'enzyme';
import * as React from 'react';
import {
    Item,
    Provider as AccordionProvider,
} from '../AccordionContainer/AccordionContainer';
import { Provider as ItemProvider } from '../ItemContainer/ItemContainer';
import { default as AccordionItemHeading } from './AccordionItemHeading.wrapper';

describe('AccordionItemHeading', () => {
    const DEFAULT_ITEM: Item = {
        uuid: 0,
        expanded: false,
    };

    it('renders null outside the context of an ‘Accordion’', () => {
        const wrapper = mount(
            <ItemProvider uuid="foo">
                <AccordionItemHeading />
            </ItemProvider>,
        );

        expect(wrapper.html()).toBeNull();
    });

    it('renders null outside the context of an ‘AccordionItem’', () => {
        const wrapper = mount(
            <AccordionProvider items={[DEFAULT_ITEM]}>
                <AccordionItemHeading />
            </AccordionProvider>,
        );

        expect(wrapper.html()).toBeNull();
    });

    function mountItem(
        node: React.ReactNode,
        item: Item = DEFAULT_ITEM,
    ): ReactWrapper {
        return mount(
            <AccordionProvider allowMultipleExpanded={true} items={[item]}>
                <ItemProvider uuid={item.uuid}>{node}</ItemProvider>
            </AccordionProvider>,
        );
    }

    function isExpanded(wrapper: ReactWrapper, uuid: string | number): boolean {
        const instance = wrapper
            .find(AccordionProvider)
            .instance() as AccordionProvider;

        return !!instance.state.items.find((item: Item) => item.uuid === uuid)
            .expanded;
    }

    it('renders correctly with min params', () => {
        const wrapper = mountItem(
            <AccordionItemHeading>
                <div>Fake Title</div>
            </AccordionItemHeading>,
        );
        expect(wrapper).toMatchSnapshot();
    });

    it('renders correctly with different className', () => {
        const className = 'className';
        const wrapper = mountItem(
            <AccordionItemHeading className={className} />,
        );
        expect(wrapper.find('div').hasClass(className)).toEqual(true);
    });

    it('renders with different expandedClassName when item is expanded', () => {
        const expandedClassName = 'expandedClassName';

        const wrapper = mount(
            <AccordionProvider items={[{ uuid: 0, expanded: true }]}>
                <ItemProvider uuid={0}>
                    <AccordionItemHeading
                        expandedClassName={expandedClassName}
                    />
                </ItemProvider>
            </AccordionProvider>,
        );
        expect(wrapper.find('div').hasClass(expandedClassName)).toEqual(true);
    });

    it('toggles state when clicked', async () => {
        const wrapper = mountItem(
            <AccordionItemHeading>Fake Title</AccordionItemHeading>,
        );

        expect(isExpanded(wrapper, DEFAULT_ITEM.uuid)).toBeFalsy();
        wrapper.find('div').simulate('click');
        expect(isExpanded(wrapper, DEFAULT_ITEM.uuid)).toBeTruthy();
    });

    it('toggles state when pressing enter', async () => {
        const wrapper = mountItem(
            <AccordionItemHeading>Fake Title</AccordionItemHeading>,
        );

        expect(isExpanded(wrapper, DEFAULT_ITEM.uuid)).toBeFalsy();
        wrapper.find('div').simulate('keyDown', { which: 13 });
        expect(isExpanded(wrapper, DEFAULT_ITEM.uuid)).toBeTruthy();
    });

    it('toggles state when pressing space', async () => {
        const wrapper = mountItem(
            <AccordionItemHeading>Fake Title</AccordionItemHeading>,
        );

        expect(isExpanded(wrapper, DEFAULT_ITEM.uuid)).toBeFalsy();
        wrapper.find('div').simulate('keyDown', { which: 32 });
        expect(isExpanded(wrapper, DEFAULT_ITEM.uuid)).toBeTruthy();
    });

    it('doesn’t toggle state when pressing another key', async () => {
        const wrapper = mountItem(
            <AccordionItemHeading>Fake Title</AccordionItemHeading>,
        );

        expect(isExpanded(wrapper, DEFAULT_ITEM.uuid)).toBeFalsy();
        wrapper.find('div').simulate('keyDown', { which: 8 });
        expect(isExpanded(wrapper, DEFAULT_ITEM.uuid)).toBeFalsy();
    });

    it('respects arbitrary user-defined props', () => {
        const wrapper = mountItem(
            <AccordionItemHeading lang="en">Fake Title</AccordionItemHeading>,
        );

        const div = wrapper.find('div').getDOMNode();

        expect(div.getAttribute('lang')).toEqual('en');
    });
});
