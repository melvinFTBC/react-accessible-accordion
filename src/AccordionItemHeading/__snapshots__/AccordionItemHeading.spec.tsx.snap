// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccordionItemHeading renders correctly with min params 1`] = `
<div
  aria-controls="accordion__panel-0"
  aria-disabled={false}
  aria-expanded={false}
  className="accordion__heading"
  data-aria-control={true}
  id="accordion__heading-0"
  onClick={[Function]}
  onKeyDown={[Function]}
  role="button"
  tabIndex={0}
>
  <div>
    Fake Title
  </div>
</div>
`;
